docs/test-faster.md

```diff
@@ -0,0 +1,191 @@
+non-test targets, such as examples or benchmarks.
+
+In an ideal world, cargo could run your tests as soon as it finished with the
+dependencies it needs for those tests, instead of waiting for your benchmark
+suite, or the arg-parser your examples use, or etc.
+
+Unfortunately, all cargo knows that these are `dev-dependencies`, and not which
+targets actually use them. (Aside: Now that per-target feature selection is
+stable we actually could fix this, however it would be extremely tedious, far
+worse than the approach outlined in this document).
+
+Additionally, unqualified invocations of cargo (that is, without `-p`) might
+have an even worse time if we aren't careful. If I run, `cargo test`, cargo
+knows *every* crate in the workspace needs to be built with all dev
+dependencies, if `places` depends on `fxa-client`, all of `fxa-clients`
+dev-dependencies must be compiled, ready, and linked in at least to the `lib`
+target before we can even think about starting on `places`.
+
+This should all sound somewhat obvious, and I guess it is. We have not been
+careful about what shape the dependency graph ends up as when example code is
+taken into consideration (as it is by cargo during certain builds), and as a
+result, we have this problem. The problem is that this isn't really a problem we
+want to fix: Example code can and should depend on several different components,
+and use them together in intersting ways.
+
+So, because we don't really want to change the things our examples do, or make
+major architectural changes of the non-test code for something like this, we
+need to do something.
+
+## The Solution
+
+To fix this, we manually insert "cuts" into the dependency graph to help cargo
+out. That is, we pull some of these build targets (e.g. examples, benchmarks,
+tests if they cause a substantial compile overhead) into their dedicated crates
+so that:
+
+1. They can be built in parallel with eachother.
+2. Crates depending on the component itself are not waiting on the
+   test/bench/example build in order for their test build to begin.
+3. A potentially smaller set of our crates need to be rebuilt -- and a smaller
+   set of possible configurations exist meaning fewer items to add pressure to
+   caches.
+4. ...
+
+There's no firm rule for this, but the following guidelines seem easy to follow:
+
+Basically, some rules of thumb for when / when not to do this:
+
+- All rust examples should be put in `examples/*`.
+
+- All rust benchmarks should be put in `testing/separated/*`. See the section
+  below on how to set your benchmark up to avoid redundant compiles.
+
+- Rust tests which brings in heavyweight dependencies should be evaluated on an
+  ad-hoc basis. If you're concerned, measure how long compilation takes
+  with/without, and consider how many crates depend on the crate where the test
+  lives (e.g. a slow test in support/foo might be far worse than one in a leaf
+  crate), etc...
```

Comment:
I'm reminded of the recent work @grigoryk did to make a [Fenix CI status dashboard](https://sql.telemetry.mozilla.org/dashboard/fenix-ci-status?p_w57980_time_period=3&p_w57980_with_tasks=OR%20job_type.symbol%20%3D%20%27gradle-dependencies%27%20OR%20job_type.symbol%20%3D%20%27ui-test-x86%27&p_w57985_time_period=3&p_w57985_with_ui_tests=OR%20job_type.symbol%20%3D%20%27gradle-dependencies%27%20OR%20job_type.symbol%20%3D%20%27ui-test-x86%27&p_w57988_time_period=3&p_w57988_with_tasks=OR%20job_type.symbol%20%3D%20%27ui-test-x86%27&p_w57990_time_period=3&p_w57990_with_tasks=OR%20job_type.symbol%20%3D%20%27gradle-dependencies%27%20OR%20job_type.symbol%20%3D%20%27ui-test-x86%27&p_w57991_time_period=3&p_w57991_with_tasks=OR%20job_type.symbol%20%3D%20%27ui-test-x86%27&p_w58003_task=T&p_w58003_time_period=3&p_w58004_task=ui-test-x86&p_w58004_time_period=3&p_w58006_task=gradle-dependencies&p_w58006_time_period=3&p_w58007_task=Bat&p_w58007_time_period=3&refresh=43200). Since you've put all this work into getting the build time down, is it worth us considering a little more work to help *keep* it down?

I don't know how involved it was to produce the above dashboard, but we might be able to borrow some or all of it from Fenix's setup.

Reply:
Filed a followup